File: ADTs/Python/Queue_test.py
from Queue import Queue
import random as rnd

Q = Queue()

assert Q.peek() == None
assert Q.isEmpty()
assert len(Q) == 0
assert not Queue().enqueue('foo').isEmpty()
assert Queue().enqueue('foo').dequeue().isEmpty()
assert Q.dequeue() == None
assert Q.enqueue('foo').peek() == 'foo'
assert Q.copy().peek() == Q.copy().enqueue('bar').peek()

L = Q.toList()
for i in range(100):
	x = rnd.randint(0, 100_000)
	Qc = Q.copy()
	Q.enqueue(x)
	assert not Q.isEmpty()
	assert len(Qc) + 1 == len(Q)
	assert Q.peek() == 'foo'
	if len(Qc) > 0:
		assert Qc.peek() == Q.peek()
		assert Qc.copy().enqueue(x).dequeue().peek() == Qc.dequeue().enqueue(x).peek()
	L.append(x)

# print(Q)
# print(L)

for i in range(len(L)):
	assert L[i] == Q.peek()
	Qc = Q.copy()
	Q.dequeue()
	assert len(Qc) - 1 == len(Q)

print('Test successful')

File: ADTs/JavaScript/Queue_test.js
import Queue from './Queue.js';

let Q = new Queue();

console.assert(Q.peek() === null, 'Peeking empty Queue should be null');
console.assert(Q.isEmpty(), 'Empty Queue should be empty');
console.assert(Q.length === 0, 'Empty Queue should have length of 0');
console.assert(!new Queue().enqueue('foo').isEmpty(), "Queue shouldn't be empty after adding element");
console.assert(new Queue().enqueue('foo').dequeue().isEmpty(), 'Enqueue then Dequeue should make new Queue empty again');
console.assert(Q.dequeue() === null, 'Dequeuing empty Queue should be null');
console.assert(Q.enqueue('foo').peek() === 'foo', 'Peeking at Q should be "foo"');
console.assert(Q.copy().peek() === Q.enqueue('bar').peek(), 'Error');

let L = Q.toList();
for (let i = 0; i < 10; i++) {
	let x = Math.floor(Math.random() * 100_000);
	let Qc = Q.copy();
	Q.enqueue(x);
	console.assert(!Q.isEmpty(), "Q shouldn't be empty");
	console.assert(Qc.length + 1 === Q.length, "Q.length didn't increase correctly");
	console.assert(Q.peek() === 'foo', "Peeking doesn't work");
	if (Qc.length) {
		console.assert(Qc.peek() === Q.peek(), 'Qc and Q should have the same first element');
		console.assert(Qc.copy().enqueue(x).dequeue().peek() === Qc.dequeue().enqueue(x).peek(), "Order of enqueuing and dequeuing shouldn't matter");
	}
	L.push(x);
}

// console.log(Q.toString());
// console.log(L);

for (let i = 0; i >= L.length; i++) {
	console.assert(L[i] === Q.peek(), 'L[i] should be equal to Q.peek(): ' + L[i] + ' and ' + Q.peek());
	let Qc = Q.copy();
	Q.dequeue();
	console.assert(Qc.length === Q.length + 1, 'Qc.length should be equal to Q.length + 1');
}

console.log('Test successful');
